### PR TITLE
Updated to Rx 4.1.0

### PR DESCRIPTION
**Pull request by ChrisPulman**
Updated projects to use Reactive 4.1.0 and latest version of ReactiveUI, added readme file to install to prompt for setup of sqlite

**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Updates System.Reactive to V4.1.0, Updates ReactiveUI to V8.7.2
Added readme file for instruction for setting up sqlite

**What is the current behavior? (You can also link to an open issue here)**
When targeting System.Reactive 4.1.0 using Akavache V6.0.20 Akavache often fails to load System.Reactive v4.0.0 if System.Reactive is not directly added to the project

**What is the new behavior (if this is a feature change)?**
Updated to V4.1.0 to remove conflict in nuget package resolution

**Does this PR introduce a breaking change?**
Updated Version of System.Reactive

**Please check if the PR fulfills these requirements**
- [x ] The commit follows our guidelines: https://github.com/Akavache/Akavache/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Other information**:

### CODE REVIEW

**Comment by glennawatson** [comment]
I think @PureWeen had some issues with Rx 4.1 and the Skip() method.

File: src/Akavache/Akavache.csproj [hunk 1]
```diff
@@ -1,15 +1,32 @@
...
     <IncludeBuildOutput>false</IncludeBuildOutput>
   </PropertyGroup>
 
+  <ItemGroup>
+    <None Remove="readme.txt" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <Content Include="readme.txt">
+      <PackagePath>\</PackagePath>
```
Comment (glennawatson) [inline src/Akavache/Akavache.csproj]: Again with the None/Content. Bit weird.

File: src/Akavache.Core/Akavache.Core.csproj [hunk 2]
```diff
@@ -6,12 +6,15 @@
     <Description>An asynchronous, persistent key-value store for desktop and mobile applications on .NET</Description>
     <PackageId>akavache.core</PackageId>
   </PropertyGroup>
+  <PropertyGroup Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Release|netstandard2.0|AnyCPU'">
+    <DocumentationFile>D:\GitHub\Akavache\src\Akavache.Core\Akavache.xml</DocumentationFile>
```
Comment (glennawatson) [inline src/Akavache.Core/Akavache.Core.csproj]: That's a very specific directory. If you do `<GenerateDocumentationFile>true</GenerateDocumentationFile>` that will output one to the correct direcotory.

File: src/Akavache.Sqlite3/Akavache.Sqlite3.csproj [hunk 3]
```diff
@@ -15,25 +15,33 @@
   <PropertyGroup Condition=" '$(Configuration)' == 'Debug' ">
     <DefineConstants>$(DefineConstants);NETFX_CORE;USE_SQLITEPCL_RAW;USE_NEW_REFLECTION_API</DefineConstants>
   </PropertyGroup>
-  
+
+  <ItemGroup>
+    <Compile Remove="%25UserProfile%25\**" />
```
Comment (glennawatson) [inline src/Akavache.Sqlite3/Akavache.Sqlite3.csproj]: What's this one about?

File: src/Akavache.Sqlite3/Akavache.Sqlite3.csproj [hunk 4]
```diff
@@ -15,25 +15,33 @@
...
   </ItemGroup>
-  
+
   <ItemGroup>
     <Compile Remove="EncryptedBlobCache.cs" />
   </ItemGroup>
 
   <ItemGroup>
     <None Remove="PerformanceGraphs.xlsx" />
+    <None Remove="readme.txt" />
```
Comment (glennawatson) [inline src/Akavache.Sqlite3/Akavache.Sqlite3.csproj]: You put it in the None section, then Content, why not just do Content once?

File: src/Akavache/Akavache.csproj (hunk 1 above)
Comment (ChrisPulman) [reply src/Akavache/Akavache.csproj]: @glennawatson was this in the test project?
[Fact(Skip = "TestScheduler tests aren't gonna work with new SQLite")]
All builds, All Tests except for those marked with the Skip attribute Pass
I have tested it in a few of my production Apps and they worked as expected.

The readme file is there to show the consumer what action is required to get the e_sqlite3.dll files in the correct place when using PackageReference. This file shows in VS when you install the package

Comment (glennawatson) [reply src/Akavache/Akavache.csproj]: Yes the point is more you have None, then later include it in Content, you can just put it in Conent once (sorry the GitHub stripped out my own review comments when I did this one.)

File: src/Akavache.Core/BitmapImageExtensions.cs [hunk 5]
```diff
@@ -19,9 +19,14 @@ public static class BitmapImageMixin
         /// <summary>
         /// Load an image from the blob cache.
         /// </summary>
+        /// <param name="This">The this.</param>
         /// <param name="key">The key to look up in the cache.</param>
-        /// <returns>A Future result representing the bitmap image. This
-        /// Observable is guaranteed to be returned on the UI thread.</returns>
+        /// <param name="desiredWidth">Width of the desired.</param>
+        /// <param name="desiredHeight">Height of the desired.</param>
```
Comment (glennawatson) [inline src/Akavache.Core/BitmapImageExtensions.cs]: You really should move all the comments out into a separate PR. Just cherry pick it out.

Eg it's not "Updating Rx 4.1" as the PR title suggests.

File: src/Akavache.Core/Akavache.Core.csproj (hunk 2 above)
Comment (ChrisPulman) [reply src/Akavache.Core/Akavache.Core.csproj]: Thanks I had missed that very specific DIr, well spotted.
I will update now

File: src/Akavache.Core/BitmapImageExtensions.cs (hunk 5 above)
Comment (glennawatson) [reply src/Akavache.Core/BitmapImageExtensions.cs]: Given you separated into different commits will be a easy one.

File: src/Akavache.Sqlite3/Akavache.Sqlite3.csproj (hunk 4 above)
Comment (ChrisPulman) [reply src/Akavache.Sqlite3/Akavache.Sqlite3.csproj]: This is the way that the new project system works as far as I know, you remove it from one group and add it to another as all files are in the none group by default

Comment (glennawatson) [reply src/Akavache.Sqlite3/Akavache.Sqlite3.csproj]: No, you can just put it into `<Content Include="readme.txt" />` once, you don't need to duplicate it.

File: src/Akavache.Core/BitmapImageExtensions.cs (hunk 5 above)
Comment (ChrisPulman) [reply src/Akavache.Core/BitmapImageExtensions.cs]: Sorry I forgot to create a new branch again. I am working through all the files in the solutions so I will ensure I branch for the next set

Comment (glennawatson) [reply src/Akavache.Core/BitmapImageExtensions.cs]: Do you want to keep all the comments in this PR? eg I can go through and validate them if you do.

File: src/Akavache.Core/BlobCache.cs [hunk 6]
```diff
@@ -130,36 +141,35 @@ public static ISecureBlobCache Secure
...
+        /// cref="DateTimeKind.Local"/> and <see cref="BsonWriter"/> uses <see
+        /// cref="DateTimeKind.Utc"/>. Thus, DateTimes are serialized as UTC but deserialized as
+        /// local time. To force BSON readers to use some other <c>DateTimeKind</c>, you can set this value.
         /// </para>
         /// </remarks>
         public static DateTimeKind? ForcedDateTimeKind { get; set; }
 
         /// <summary>
-        /// 
+        /// ///
```
Comment (glennawatson) [inline src/Akavache.Core/BlobCache.cs]: May want to make a actual comment here, rather than  just adding `///`

File: src/Akavache.Core/BulkOperationsMixin.cs [hunk 7]
```diff
@@ -10,80 +10,133 @@ namespace Akavache
...
+                return bulkCache.Insert(keyValuePairs, absoluteExpiration);
+            }
 
             return keyValuePairs.ToObservable()
                 .SelectMany(x => This.Insert(x.Key, x.Value, absoluteExpiration))
                 .TakeLast(1);
         }
 
+        /// <summary>
+        /// Gets the created at.
```
Comment (glennawatson) [inline src/Akavache.Core/BulkOperationsMixin.cs]: Maybe something like "Get the created date time of the specified blob cache key entries"

File: src/Akavache.Core/BulkOperationsMixin.cs [hunk 8]
```diff
@@ -10,80 +10,133 @@ namespace Akavache
...
+                return bulkCache.Invalidate(keys);
+            }
 
             return keys.ToObservable()
                .SelectMany(x => This.Invalidate(x))
                .TakeLast(1);
         }
 
+        /// <summary>
+        /// Gets the objects.
```
Comment (glennawatson) [inline src/Akavache.Core/BulkOperationsMixin.cs]: "Gets the objects with the specified keys"

File: src/Akavache.Core/BulkOperationsMixin.cs [hunk 9]
```diff
@@ -10,80 +10,133 @@ namespace Akavache
...
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
```
Comment (glennawatson) [inline src/Akavache.Core/BulkOperationsMixin.cs]: The type of the stored entries.

File: src/Akavache.Core/BulkOperationsMixin.cs [hunk 10]
```diff
@@ -10,80 +10,133 @@ namespace Akavache
...
                         .Catch<KeyValuePair<string, T>, KeyNotFoundException>(_ => Observable.Empty<KeyValuePair<string, T>>());
                 })
-                .ToDictionary(k => k.Key, v => v.Value);           
+                .ToDictionary(k => k.Key, v => v.Value);
         }
 
+        /// <summary>
+        /// Inserts the objects.
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
```
Comment (glennawatson) [inline src/Akavache.Core/BulkOperationsMixin.cs]: The type of the objects to be inserted.

File: src/Akavache.Core/IEncryptionProvider.cs [hunk 11]
```diff
@@ -8,8 +8,18 @@ namespace Akavache
 {
     public interface IEncryptionProvider
     {
+        /// <summary>
+        /// Encrypts the block.
+        /// </summary>
+        /// <param name="block">The block.</param>
+        /// <returns></returns>
```
Comment (glennawatson) [inline src/Akavache.Core/IEncryptionProvider.cs]: Provide a returns?

File: src/Akavache.Core/IFilesystemProvider.cs [hunk 12]
```diff
@@ -60,44 +57,100 @@ public interface IFilesystemProvider
...
     }
-
 }
 
 namespace Akavache.Internal
 {
     public enum FileMode
     {
+        /// <summary>
+        /// The create new
```
Comment (glennawatson) [inline src/Akavache.Core/IFilesystemProvider.cs]: I get the feeling these are very auto-generated. Not that useful really.

File: src/Akavache.Core/BitmapImageExtensions.cs (hunk 5 above)
Comment (ChrisPulman) [reply src/Akavache.Core/BitmapImageExtensions.cs]: If you have the time to review it, then it would be nice to get this one in too, I am sorry there's many changes mostly due to the code style change between spaces and tabs

File: src/Akavache.Core/InMemoryBlobCache.cs [hunk 13]
```diff
@@ -264,38 +393,39 @@ byte[] SerializeObject<T>(T value)
...
-            {
+            } catch (Exception ex) {
                 this.Log().WarnException("Failed to deserialize data as boxed, we may be migrating from an old Akavache", ex);
             }
 
             return serializer.Deserialize<T>(reader);
         }
 
+        /// <summary>
+        /// Overrides the globals.
```
Comment (glennawatson) [inline src/Akavache.Core/InMemoryBlobCache.cs]: Something a bit more meaningful?

File: src/Akavache.Core/InMemoryBlobCache.cs [hunk 14]
```diff
@@ -326,11 +468,37 @@ public static InMemoryBlobCache OverrideGlobals(IDictionary<string, object> init
 
     public class CacheEntry
     {
+        /// <summary>
+        /// Gets or sets the created at.
```
Comment (glennawatson) [inline src/Akavache.Core/InMemoryBlobCache.cs]: Bit more detail than the auto-generated comments here be useful.

File: src/Akavache.Core/JsonSerializationMixin.cs [hunk 15]
```diff
@@ -7,78 +8,89 @@
...
             return This.Insert(GetTypePrefixedKey(key, typeof(T)), bytes, absoluteExpiration);
         }
 
         /// <summary>
-        /// Insert several objects into the cache, via the JSON serializer. 
-        /// Similarly to InsertAll, partial inserts should not happen.
+        /// Insert several objects into the cache, via the JSON serializer. Similarly to InsertAll,
+        /// partial inserts should not happen.
         /// </summary>
+        /// <typeparam name="T"></typeparam>
```
Comment (glennawatson) [inline src/Akavache.Core/JsonSerializationMixin.cs]: Bunch of empty typeparam in this file.

File: src/Akavache.Core/KeyedOperationQueue.cs [hunk 16]
```diff
@@ -51,62 +58,72 @@ public KeyedOperationQueue(IScheduler scheduler = null)
...
             };
 
             queuedOps.OnNext(item);
             return item.Result;
         }
 
+        /// <summary>
+        /// Flushes the remaining operations and returns a signal when they are all complete.
+        /// </summary>
+        /// <returns></returns>
```
Comment (glennawatson) [inline src/Akavache.Core/KeyedOperationQueue.cs]: Empty returns.

File: src/Akavache.Core/Platforms/android/AndroidFilesystemProvider.cs [hunk 17]
```diff
@@ -10,46 +10,81 @@ namespace Akavache
...
         public IObservable<Stream> OpenFileForWriteAsync(string path, IScheduler scheduler)
         {
             return _inner.OpenFileForWriteAsync(path, scheduler);
         }
 
+        /// <summary>
+        /// Creates the recursive.
+        /// </summary>
+        /// <param name="path">The path.</param>
+        /// <returns></returns>
```
Comment (glennawatson) [inline src/Akavache.Core/Platforms/android/AndroidFilesystemProvider.cs]: Bunch of empty returns in this file.

File: src/Akavache.Core/Platforms/apple-common/MacFilesystemProvider.cs [hunk 18]
```diff
@@ -1,63 +1,101 @@
...
     {
-        readonly SimpleFilesystemProvider _inner = new SimpleFilesystemProvider();
+        private readonly SimpleFilesystemProvider _inner = new SimpleFilesystemProvider();
 
+        /// <summary>
+        /// Opens the file for read asynchronous.
+        /// </summary>
+        /// <param name="path">The path.</param>
+        /// <param name="scheduler">The scheduler.</param>
+        /// <returns></returns>
```
Comment (glennawatson) [inline src/Akavache.Core/Platforms/apple-common/MacFilesystemProvider.cs]: Bunch of empty returns in this file.

File: src/Akavache.Core/Platforms/shared-not-uwp/SimpleFilesystemProvider.cs [hunk 19]
```diff
@@ -1,63 +1,101 @@
...
 namespace Akavache
 {
     public class SimpleFilesystemProvider : IFilesystemProvider
     {
+        /// <summary>
+        /// Opens the file for read asynchronous.
+        /// </summary>
+        /// <param name="path">The path.</param>
+        /// <param name="scheduler">The scheduler.</param>
+        /// <returns></returns>
```
Comment (glennawatson) [inline src/Akavache.Core/Platforms/shared-not-uwp/SimpleFilesystemProvider.cs]: Bunch of empty returns in this file.

File: src/Akavache.Core/Platforms/shared/StreamMixins.cs [hunk 20]
```diff
@@ -11,36 +10,37 @@ namespace System
...
     public static class StreamMixins
     {
+        /// <summary>
+        /// Writes the asynchronous rx.
+        /// </summary>
+        /// <param name="This">The this.</param>
+        /// <param name="data">The data.</param>
+        /// <param name="start">The start.</param>
+        /// <param name="length">The length.</param>
+        /// <returns></returns>
```
Comment (glennawatson) [inline src/Akavache.Core/Platforms/shared/StreamMixins.cs]: Comment the returns.

File: src/Akavache.Core/Platforms/uap10.0.16299/WinRTEncryptionProvider.cs [hunk 21]
```diff
@@ -2,30 +2,39 @@
...
 
 namespace Akavache
 {
     public class WinRTEncryptionProvider : IEncryptionProvider
     {
+        /// <summary>
+        /// Encrypts the block.
+        /// </summary>
+        /// <param name="block">The block.</param>
+        /// <returns></returns>
```
Comment (glennawatson) [inline src/Akavache.Core/Platforms/uap10.0.16299/WinRTEncryptionProvider.cs]: Bunch of empty comment blocks.

File: src/Akavache.Sqlite3/Akavache.Sqlite3.csproj (hunk 3 above)
Comment (ChrisPulman) [reply src/Akavache.Sqlite3/Akavache.Sqlite3.csproj]: I have deleted this rather bizarre added code from VS

File: src/Akavache.Core/Platforms/xamarin-mobile/IsolatedStorageProvider.cs [hunk 22]
```diff
@@ -14,107 +14,121 @@ namespace Akavache
 {
     public class IsolatedStorageProvider : IFilesystemProvider
     {
+        /// <summary>
+        /// Opens the file for read asynchronous.
+        /// </summary>
+        /// <param name="path">The path.</param>
+        /// <param name="scheduler">The scheduler.</param>
+        /// <returns></returns>
```
Comment (glennawatson) [inline src/Akavache.Core/Platforms/xamarin-mobile/IsolatedStorageProvider.cs]: Bunch of empty returns blocks.

File: src/Akavache.Core/RelativeTimeMixin.cs [hunk 23]
```diff
@@ -5,26 +5,64 @@
...
 
+        /// <summary>
+        /// Downloads the URL.
+        /// </summary>
+        /// <param name="This">The this.</param>
+        /// <param name="url">The URL.</param>
+        /// <param name="expiration">The expiration.</param>
+        /// <param name="headers">The headers.</param>
+        /// <param name="fetchAlways">if set to <c>true</c> [fetch always].</param>
+        /// <returns></returns>
```
Comment (glennawatson) [inline src/Akavache.Core/RelativeTimeMixin.cs]: Bunch of empty returns blocks.

**Comment by glennawatson** [comment]
This is from @PureWeen on the Slack channel

> Rx 4.1 is broken on Xamarin Android. I think we should wait until that's resolved before forcing the version.

Why take a forced dependency on 4.1? Users can just use 4.1

github APP [3:33 PM]
[reactiveui/Akavache] ChrisPulman commented on #426: Updated to Rx 4.1.0
If you have the time to review it, then it would be nice to get this one in too, I am sorry there's many changes mostly due to the code style change between spaces and tabs
Comment by ChrisPulman on line 15 of src/Akavache.Core/BitmapImageExtensions.cs

Shane Neuville [3:34 PM]
https://github.com/dotnet/reactive/issues/803
GitHub
Cannot compile Xamarin Android project with System.Reactive v4.1.0 · Issue #803 · dotnet/reactive
Description The new dependency System.Threading.Tasks.Extensions cannot be fully resolved on Xamarin Android projects. This is not a problem of Rx.NET but I&#39;m leaving this here so other people ...
I'd post to the github issue but I'm on vacation and don't have access to my github account

File: src/Akavache/Akavache.csproj (hunk 1 above)
Comment (ChrisPulman) [reply src/Akavache/Akavache.csproj]: Updated as requested

File: src/Akavache.Core/BitmapImageExtensions.cs (hunk 5 above)
Comment (glennawatson) [reply src/Akavache.Core/BitmapImageExtensions.cs]: The change are good ones for the comments/code style ones in my opinion. Just a lot of generic auto-generated ones that feel a bit worthless.

**Comment by glennawatson** [comment]
https://github.com/dotnet/reactive/issues/803

**Comment by glennawatson** [comment]
As per that thread that @PureWeen  referenced you may want to do the following

> In-box, at least in the Xamarin context, means that the deployed runtime got the necessary bits already bundled, so that a referenced nuget package may just omit any files for that platform. The System.Threading.Tasks.Extensions nuget in version 4.5.1 assumes the appropriate dlls be present on the Xamarin platform (in-box!), but they aren't. It's a bug, it'll be fixed eventually, until that, downgrade to System.Threading.Tasks.Extensions 4.4.0 which is just fine.
> Out of box means the opposite: The nuget has to provide the actual dlls, which is of course the case for most stuff on nuget.

**Comment by ChrisPulman** [comment]
@glennawatson Thanks, I saw an issue with the Android test project and thought it was just out of date as many of the other ones are there, I was going to try and fix this and submit as a separate pr for the Test projects solution.
I will still submit the pr's for the code clean up but stick to Rx 4.0.0 until there is a resolve
